src/Ogooreck.Sample.BusinessLogic.FSharp.Tests/Deciders/BankAccountDecider.fs

```diff
@@ -0,0 +1,80 @@
+﻿module Deciders.BankAccountDecider
+
+open System
+open BankAccount
+open BankAccountPrimitives
+
+type OpenBankAccount =
+    { BankAccountId: AccountId
+      AccountNumber: AccountNumber
+      ClientId: ClientId
+      CurrencyISOCode: CurrencyCode
+      Now: DateTimeOffset }
+
+type RecordDeposit = { Amount: decimal; CashierId: CashierId; Now: DateTimeOffset }
+type WithdrawCashFromATM = { Amount: decimal; AtmId: AtmId; Now: DateTimeOffset }
+type CloseBankAccount = { Reason: string; Now: DateTimeOffset }
+
+type Command =
+    | OpenBankAccount of OpenBankAccount
+    | RecordDeposit of RecordDeposit
+    | WithdrawCashFromATM of WithdrawCashFromATM
+    | CloseBankAccount of CloseBankAccount
+
+let openBankAccount (command: OpenBankAccount) (bankAccount: BankAccount) =
+    match bankAccount with
+    | Open _ -> invalidOp "Account is already opened!"
```

Comment:
idempotency FTW

Reply:
Consider having this function return a `Result` rather than `invalidOp`?

Reply:
Problem is that the Result then becomes a union of all the possible result types for all Command DU cases. (but really, idempotency FTW - you need to be able to handle retries without fprcing people to ignore exceptions they can't prevent)

Reply:
I didn't expect this to happen, and I didn't intend to recover, so I used exceptions. In other cases, I'd use results. Regarding idempotency, if I ignore the Open case here, then I'll need to return `None` and make the proper case return also Option, right? By that, I'm saying for this user that the case where no event is returned may happen, and they should handle it. When I'm treating this case as exceptional, I don't want to force the user to handle that each time they call the method and propagate that to other places. So IMHO here implicitness is more explicit 😉 
Thoughts?

Reply:
> if I ignore the Open case here, then I'll need to return None and make the proper case return also Option, right?

`list` is the ideal way to represent the events - allows 0, 1 or many naturally. List comprehension syntax makes it nice and clean.

(if the caller needs a result of some kind too, I say that's orthogonal and you should return a tuple combining the result and the event[list])

(I've been around the houses with this myself in C# and F# with every possible combination and am confident seq/list is the answer here and am  confident you'll eventually come around too. In the meantime, you'll learn plenty from making things return a single event or an `option` - my main problem is that dumbing down the example by not bringing idempotency to the fore just so you don't need to deal with 0/1/many events is not really helping anyone - you don't want them to have to go way off your happy path when they learn about idempotency (and/or they come wanting to deal with it already))

----
 
And I agree with throws for guards for preventable stuff over egregious use of Result wherever possible.